website/addons/googledrive/model.py

```diff
@@ -111,218 +101,183 @@ def get_or_create(cls, node, path):
-        # if there is only one osf user linked to this google drive user oauth, revoke the token,
-        # otherwise, disconnect the osf user from the googledriveoauthsettings
-        if len(self.googledriveusersettings__accessed) <= 1:
-            client = GoogleAuthClient()
-            try:
-                client.revoke(self.access_token)
-            except:
-                # no need to fail, revoke is opportunistic
-                pass
-
-            # remove the object as its the last instance.
-            GoogleDriveOAuthSettings.remove_one(self)
+            token = self._refresh_token(self.account.oauth_key, self.account.refresh_token)
+            self.account.oauth_key = token['access_token']
+            self.account.refresh_token = token['refresh_token']
+            self.account.expires_at = datetime.utcfromtimestamp(token['expires_at'])
+            self.account.save()
 
     def _needs_refresh(self):
-        if self.expires_at is None:
+        if self.account.expires_at is None:
             return False
-        return (self.expires_at - datetime.utcnow()).total_seconds() < drive_settings.REFRESH_TIME
+        return (self.account.expires_at - datetime.utcnow()).total_seconds() < drive_settings.REFRESH_TIME
 
 
-class GoogleDriveUserSettings(AddonUserSettingsBase):
-    """Stores user-specific information, including the Oauth access
-    token.
-    """
-    oauth_settings = fields.ForeignField(
-        'googledriveoauthsettings', backref='accessed'
-    )
+class GoogleDriveUserSettings(StorageAddonBase, AddonOAuthUserSettingsBase):
+    oauth_provider = GoogleDriveProvider
+    oauth_grants = fields.DictionaryField()
+    serializer = GoogleDriveSerializer
 
-    @property
-    def user_id(self):
-        if self.oauth_settings:
-            return self.oauth_settings.user_id
-        return None
 
-    @user_id.setter
-    def user_id(self, val):
-        self.oauth_settings.user_id = val
+class GoogleDriveNodeSettings(StorageAddonBase, AddonOAuthNodeSettingsBase):
+    oauth_provider = GoogleDriveProvider
 
-    @property
-    def username(self):
-        if self.oauth_settings:
-            return self.oauth_settings.username
-        return None
+    drive_folder_id = fields.StringField(default=None)
+    drive_folder_name = fields.StringField(default=None)
+    folder_path = fields.StringField()
+    serializer = GoogleDriveSerializer
 
-    @username.setter
-    def username(self, val):
-        self.oauth_settings.username = val
+    _api = None
 
     @property
-    def access_token(self):
-        if self.oauth_settings:
-            return self.oauth_settings.access_token
-        return None
-
-    @access_token.setter
-    def access_token(self, val):
-        self.oauth_settings.access_token = val
+    def api(self):
+        """Authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = GoogleDriveProvider()
+            self._api.account = self.external_account
```

Comment:
It should, but doesn't. That isn't directly related to this PR's purpose, but I could make that change if you think it's worthwhile.